Docs: window aggregations are now supported (#22552)
MaterializeInc/materialize: doc/user/content/transform-data/patterns/window-functions.md

    parent: 'sql-patterns'
---

Materialize offers a wide range of [window functions](/sql/functions/#window-func), such as `LAG`, `LEAD`, `ROW_NUMBER`, and `FIRST_VALUE`, to support a variety of complex query patterns. This page discusses certain window function query patterns where rewriting your query to not use window functions can yield a better performance.

Currently, Materialize implements window functions by recomputing results for an entire window partition when an input record is added, removed or changed in the partition. This means that when a new batch of input data arrives (that is, every second) the system has to perform an amount of computation proportional to the total size of the touched partitions (where the partitions are determined by the `PARTITION BY` clause inside the `OVER` clause of your window function call). For example, if in a certain second 20 input records change, and these records belong to 10 different partitions, and the average sizes of these partitions is 100, then the system will perform work proportional to computing window function results for 10*100=1000 rows.

> As a rule of thumb, if the total size of all touched window partitions is at most 1000000 rows per second, than the system will be able to keep up with the input data as it arrives in most cases. However, if you find that your use case has higher performance requirements, then consider rewriting your query to not use window functions. This page helps with such rewrites.

Reviewer: ```suggestion
As a rule of thumb, if the total size of all touched window partitions is at most 1000000 rows per second, then the system should be able to keep up with the input data as it arrives. However, if you find that your use case has higher performance requirements, then consider rewriting your query to not use window functions. This page helps with such rewrites.
```

Author: Done, thx!

    parent: 'sql-patterns'
---

> Some query patterns that are commonly used in batch systems are tricky to support with continuously arriving input data. Window functions, such as `LAG`, `LEAD`, `ROW_NUMBER`, `FIRST_VALUE`, are one of those patterns.

Reviewer 2: Feel free to rewrite, but we should open with a sentence that clearly states we do support window functions. 
```suggestion
Materialize implements a wide range of window functions to support a variety of complex query patterns. 
Some query patterns that are commonly used in batch systems are tricky to support with continuously arriving input data. Window functions, such as `LAG`, `LEAD`, `ROW_NUMBER`, `FIRST_VALUE`, are one of those patterns.
```

Author: Thanks! I made the change.
